### ISSUE TITLE
Stream-GlobalTable left join not working as expected

### ISSUE THREAD

**Issue by mayur-rao** [LGouellec/kafka-streams-dotnet#165]
Description
===========
Performing a Stream-GlobalTable left join does not pickup matching keys. Right side values (from table) are always empty/null. The same logic implemented in Java with Streams API works correctly and picks up the matching keys correctly. 

How to reproduce
================
Topic 1 : JoinSample-1
Messages --> Key:Value
100:tableValue-1
200:tableValue-2
300:tableValue-3

Topic 2 : JoinSample-2
Messages --> Key:Value
100:streamValue-4
300:streamValue-3
50:streamValue-8
500:streamValue-2
600:streamValue-1
700:streamValue-5
800:streamValue-6
900:streamValue-7

EXPECTED JoinedEvent Messages --> Key:Value
100:streamValue-4joinedWithtableValue-1
300:streamValue-3joinedWithtableValue-3
50:streamValue-8joinedWith
500:streamValue-2joinedWith
600:streamValue-1joinedWith
700:streamValue-5joinedWith
800:streamValue-6joinedWith
900:streamValue-7joinedWith

ACTUAL JoinedEvent Messages (fails to pickup values for Key 100 and 300 from Table topic) --> Key:Value
100:streamValue-4joinedWith
300:streamValue-3joinedWith
50:streamValue-8joinedWith
500:streamValue-2joinedWith
600:streamValue-1joinedWith
700:streamValue-5joinedWith
800:streamValue-6joinedWith
900:streamValue-7joinedWith

Checklist
=========

Please provide the following information:
 - [ ] A complete (i.e. we can run it), minimal program demonstrating the problem. No need to supply a project file. - code attached
 - [ ] A code snippet with your topology builder (ex: builder.Stream<string, string>("topic").to("an-another-topic");) - see attached code
 - [ ] Streamiz.Kafka.Net nuget version. - 1.3.0
 - [ ] Apache Kafka version. - 6.0.1-ce Confluent Server
 - [ ] Client configuration. --> 
```
 var config = new StreamConfig<StringSerDes, StringSerDes>
            {
                ApplicationId = $"test-kstreams-{Guid.NewGuid()}",
                BootstrapServers = "localhost:29092",
                AutoOffsetReset = AutoOffsetReset.Earliest
            };
```
 - [ ] Operating system. - linux
 - [ ] Provide logs (with in debug mode (log4net and StreamConfig.Debug) as necessary in configuration).  - attached file
 - [ ] Critical issue - NA
[Program.txt](https://github.com/LGouellec/kafka-streams-dotnet/files/9236945/Program.txt)
[logs.txt](https://github.com/LGouellec/kafka-streams-dotnet/files/9236956/logs.txt)
.

**Comment by LGouellec** [comment]
Hi @mayur-rao,

Can you confirm that both topics are messages before starting your application ?

Kr,

**Comment by mayur-rao** [comment]
Hi @LGouellec , 

Yes, I can confirm that both topics have messages before I run the streams app. Also I made sure that the timestamps for the topic I want to use as table (JoinSample-1 above) are before the topic I want to use as stream (JoinSample-2).

**Comment by LGouellec** [comment]
@mayur-rao, 

Ok I found the issue. 
It's just the GlobalStreamThread and the StreamThread are started in parallel. 
So the GlobalStreamThread restore the global state store in the same time than the stream thread process data from `JoinSample-2` topic. In your case, stream thread process messages before the global state store is restored.

That's why if you wait some seconds after the startup, if you post an another time this message into `JoinSample-2` :
300:streamValue-3

The join works fine: 300:streamValue-3joinedWithtableValue-3

I think the JAVA implementation works like : 
- Start global stream thread and restore global state store with messages present in the global topic
- Wait end of restoration for the global state store
- Start stream thread and begin to process data.

I have to check how add the step 2 in my library.

Br,

**Comment by mayur-rao** [comment]
Hi @LGouellec , 

Thanks for looking. I think I understand and will try adding a wait and inspect the behavior. As far as the fix is concerned, will it be released in 1.4.0 package version? Also can you tell me when 1.4.0 will be out?

Thanks

**Comment by LGouellec** [comment]
> Hi @LGouellec , 
> 
> 
> 
> Thanks for looking. I think I understand and will try adding a wait and inspect the behavior. As far as the fix is concerned, will it be released in 1.4.0 package version? Also can you tell me when 1.4.0 will be out?
> 
> 
> 
> Thanks

I think I will create a hot fix 1.3.1. Maybe next week or begin of September.

**Comment by LGouellec** [comment]
1.3.1 is released. So I close this issue, please open a new ticket if the problem still existing or something else.

**Comment by mayur-rao** [comment]
Wonderful. I tested and looks like it is working as expected. Thanks for the quick fix.
